pkg/kv/kvserver/replica_application_result.go

```diff
@@ -379,6 +385,27 @@ func (r *Replica) tryReproposeWithNewLeaseIndex(ctx context.Context, origCmd *re
 		}
 	}()
 
+	defer func() {
+		if !success {
+			return
+		}
+		// If the proposal is synchronous, the client is waiting on the seed
+		// proposal. By the time it has to act on the result, a bunch of reproposals
+		// can have happened, and some may still be running and using the
+		// context/tracing span.
+		//
+		// Unbind the latest reproposal's context so that it no longer posts updates
+		// to the tracing span (it won't apply anyway). Link to the new latest
+		// reproposal, so that the client can clear its context at post-processing.
+		//
+		// TODO(pavelkalinnikov): there should be a better way, after ProposalData
+		// lifecycle is reconsidered.
+		if last := seedP.lastReproposal; last != nil {
+			last.ctx = r.AnnotateCtx(context.TODO())
+		}
```

Comment:
Do we expect these abandoned proposals to ever be used, or are they discarded after we return here?

I think this code makes sense as defense-in-depth, but in principle we should be able to just leave it, right?

Reply:
Logically you're right - older proposals won't be applied so we can forget them. Physically though, I'm not sure that there isn't some kind of race in result of which an older proposal comes up and "used" (for some clean-up) even after a newer proposal has been submitted.

So yeah, this is defence in depth. Need to mark it as such in the comment, and revisit.

Reply:
I would really prefer having one context/trace shared by all (re-)proposals explicitly.

Reply:
Noted in the comment.